rfd/0030-upgrade-system.md

```diff
@@ -0,0 +1,444 @@
+a given server.
+
+Policy evaluation (i.e. determining how releases map to servers) will be the responsibility
+of the auth server.  The goal of this centralization is to ensure that if a logic bug in the
+mapping results in requiring a manual upgrade, only the auth server(s) will need to be upgraded.
+Some amount of hedging will be added to the API to make sure its easy to add custom (external)
+policy evaluators later if the need exists, but I imagine that custom upgrade controllers and
+manually managed release manifests will cover most usecases.
+
+I have yet to work out the exact details of how policy syntax and evaluation should work, but
+here's an example of one of the syntaxes I've been toying with in my head:
+
+```yaml
+target_selectors:
+  - name: Latest oss
+    flavors: [oss]
+    permit_unstable: true
+server_selectors:
+  - name: Preview servers
+    server_roles: ['proxy','node']
+    filter: 'contains(server.labels['unstable-preview'],'yes')'
+schedule:
+  not_before: '2021-04-29T00:00:00Z'
+  time_range: '01:00:00-06:00:00'
+  day_range: Mon-Fri
+```
+
+Note that my current thinking on target/server matching is that policies should not actually
+be able to cause a node to be upgraded to a target with different fundamental attributes (e.g. os,
+flavor, arch, etc) due to the risk of catastrophic user error being too great. Instead, the
+purpose of target selectors it to allow different requirements to be put in place *based* on
+those fundamental attributes (e.g. only installing unstable versions on oss nodes).
+
+Not all policies need be fancy.  In fact, something like this aught to be a valid policy:
+
+```
+target_selectors:
+- version: '99.*'
+```
+
+read: "Keep all servers on the latest stable `99.X` release"
+
+
+#### Upgrade Controllers
+
+The "upgrade controller" will be the primary hook that is used to customize upgrade
+behavior.  An upgrade controller takes in config objects, and `server => target`
+mappings produced by upgrade policies, and attempts to reconcile the current state
+of the various servers with their target states. 
+
+In a perfect world, we'd launch the upgrade system with an in-house installer *so good*
+that everyone would immediately drop their existing package management solutions and
+wonder how they ever got by before.  In reality, this is a pretty tough sell, and
+a huge amount of work if you want to build something meaningfully better than can
+be achieved with a few lines of bash.  With this in mind, I'm of the opinion that
+flexibility of upgrade mechanism matters more than having a single extremely robust
+install mechanism, at least where MVP functionality is concerned.
+
+Lets take a look at a hypothetical example of an input resource for a "local script"
+upgrade controller:
+
+```yaml
+env:
+  TELEPORT_VER: '${target.version}'
+  TELEPORT_SUM: '${target.sums["rpm-blake2-256"]}'
+install.sh: |
+  #!/bin/bash
+  set -euo pipefail
```

Comment:
I think @knisbet will appreciate the pluggable controllers with configurable steps. I feel like we have incorporated his feedback here with both controller, convergence and scripting :) One of the biggest problems with  our gravity's upgrade system was that everything was in Go and it was impossible for customers to troubleshoot failed steps or reconfigure them. The pluggable controller + script solves that problem.